index.js

```diff
@@ -3757,6 +3757,51 @@
   }
   S.singleton = def('singleton', {}, [$.String, a, $.StrMap(a)], singleton);
 
+  //# insert :: String -> a -> StrMap a -> StrMap a
+  //.
+  //. Takes a string, a value of any type, and a string map, and returns a
+  //. string map comprising all the entries of the given string map plus the
+  //. entry specified by the first two arguments (which takes precedence).
+  //.
+  //. Equivalent to Haskell's `insert` function. Similar to Clojure's `assoc`
+  //. function.
+  //.
+  //. ```javascript
+  //. > S.insert('c', 3, {a: 1, b: 2})
+  //. {a: 1, b: 2, c: 3}
+  //.
+  //. > S.insert('a', 4, {a: 1, b: 2})
+  //. {a: 4, b: 2}
+  //. ```
+  function insert(key, val, strMap) {
+    return Z.concat(strMap, singleton(key, val));
```

Comment:
I like this implementation in particular 😄

Reply:
Is there a policy of not "cannibalizing" Sanctuary functions to implement other things? You could do something like `const insert = compose2(concat, singleton)`. I guess without a type system to guide you getting too abstract gets confusing.

Reply:
I think `insert`\`assoc` is used frequently enough to rate it's own function. Unless I'm misunderstanding you.

Reply:
@gabejohnson I agree with having `insert` as a separate function, I didn't mean to say this shouldn't be part of the library. I'm just saying that when we implement these functions, it seems like we prefer to use `Z` in the implementation even when it makes the implementation a little more awkward than if we used the equivalent `S` function.

Reply:
This is because of the overhead introduced by `def` (currying, redundant type checking machinery). _Some_ care is taken to write performant implementations.

Reply:
Haha! Your use of italics just made me chuckle, Gabe. :)

We avoid using the curried functions, Asad, for the reason Gabe gave, but the *implementation* functions can freely be used in other implementations. The `head` implementation, for example, is defined in terms of the `at` implementation:

> ```javascript
> function head(xs) {
>   return at(0, xs);
> }
> ```